### ISSUE TITLE
Code: 1001 DB::Exception filesystem_error No such file or directory on Queries

### ISSUE THREAD

**Issue by filipealc** [ClickHouse/ClickHouse#68344]
#### Clickhouse Version
- 24.1.5.6

**Company or project name**

- VTurb

**Describe what's wrong**

Queries (select and insert) randomly fail with `DB::ErrnoException: Cannot open file...` on tables with projections while using a distributed cluster with s3 as storage.

**Full log**

```text
ClickHouse::DbException ([200] {"meta"=>[{"name"=>"company_id", "type"=>"String"}, {"name"=>"event", "type"=>"LowCardinality(String)"}, {"name"=>"total", "type"=>"Int64"}, {"name"=>"total_uniq_sessions", "type"=>"Int64"}, {"name"=>"total_uniq_device", "type"=>"Int64"}], "data"=>[], "rows"=>0, "exception"=>"Code: 107. DB::Exception: Received from [REDACTED]:9000. DB::ErrnoException. DB::ErrnoException: Cannot open file /var/lib/clickhouse/store/216/21607f49-9c9a-42d5-8a84-7ec971117969/202408_3885_3885_0/general_view_by_event.proj/data.bin: , errno: 2, strerror: No such file or directory: While executing MergeTreeSelect(pool: ReadPool, algorithm: Thread). Stack trace:
0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000bc1a1e8 in /usr/bin/clickhouse
1. DB::ErrnoException::ErrnoException(String&&, int, int) @ 0x000000000bc53a1c in /usr/bin/clickhouse
2. void DB::ErrnoException::throwFromPath<String const&>(int, String const&, FormatStringHelperImpl<std::type_identity<String const&>::type>, String const&) @ 0x000000000bc53650 in /usr/bin/clickhouse
3. DB::OpenedFile::getFD() const @ 0x000000000ee5898c in /usr/bin/clickhouse
4. DB::AsynchronousReadBufferFromFileWithDescriptorsCache::AsynchronousReadBufferFromFileWithDescriptorsCache(DB::IAsynchronousReader&, Priority, String const&, unsigned long, int, char*, unsigned long, std::optional<unsigned long>, std::shared_ptr<DB::Throttler>) @ 0x000000000ee57f54 in /usr/bin/clickhouse
5. DB::createReadBufferFromFileBase(String const&, DB::ReadSettings const&, std::optional<unsigned long>, std::optional<unsigned long>, int, char*, unsigned long)::$_0::operator()(unsigned long, unsigned long, int) const @ 0x000000000ee550fc in /usr/bin/clickhouse
6. DB::createReadBufferFromFileBase(String const&, DB::ReadSettings const&, std::optional<unsigned long>, std::optional<unsigned long>, int, char*, unsigned long) @ 0x000000000ee54dc0 in /usr/bin/clickhouse
7. DB::DiskLocal::readFile(String const&, DB::ReadSettings const&, std::optional<unsigned long>, std::optional<unsigned long>) const @ 0x000000000f7bfc08 in /usr/bin/clickhouse
8. DB::DataPartStorageOnDiskFull::readFile(String const&, DB::ReadSettings const&, std::optional<unsigned long>, std::optional<unsigned long>) const @ 0x0000000010c0633c in /usr/bin/clickhouse
9. DB::MergeTreeReaderCompact::initialize() @ 0x0000000010df5874 in /usr/bin/clickhouse
10. DB::MergeTreeReaderCompact::readRows(unsigned long, unsigned long, bool, unsigned long, std::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn>>>&) @ 0x0000000010df64a8 in /usr/bin/clickhouse
11. DB::MergeTreeRangeReader::DelayedStream::finalize(std::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn>>>&) @ 0x0000000010e15fdc in /usr/bin/clickhouse
12. DB::MergeTreeRangeReader::read(unsigned long, DB::MarkRanges&) @ 0x0000000010e1eaec in /usr/bin/clickhouse
13. DB::MergeTreeReadTask::read(DB::MergeTreeReadTask::BlockSizeParams const&) @ 0x0000000010e257b4 in /usr/bin/clickhouse
14. DB::MergeTreeSelectProcessor::read() @ 0x0000000010e13e70 in /usr/bin/clickhouse
15. DB::MergeTreeSource::tryGenerate() @ 0x00000000115a722c in /usr/bin/clickhouse
16. DB::ISource::work() @ 0x00000000111a64b4 in /usr/bin/clickhouse
17. DB::ExecutionThreadContext::executeTask() @ 0x00000000111ba96c in /usr/bin/clickhouse
18. DB::PipelineExecutor::executeStepImpl(unsigned long, std::atomic<bool>*) @ 0x00000000111b2af4 in /usr/bin/clickhouse
19. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<DB::PipelineExecutor::spawnThreads()::$_0, void ()>>(std::__function::__policy_storage const*) @ 0x00000000111b3c7c in /usr/bin/clickhouse
20. ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) @ 0x000000000bce7030 in /usr/bin/clickhouse
21. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x000000000bcea220 in /usr/bin/clickhouse
22. void* std::__thread_proxy[abi:v15000]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000000bce91e8 in /usr/bin/clickhouse
23. ? @ 0x0000ffff8624c834
24. ? @ 0x0000ffff861f0e5c
: While executing Remote. (FILE_DOESNT_EXIST) (version 24.1.5.6 (official build))"})
```

Random error on INSERTs

```text
Code: 1001. DB::Exception: Received from ch-server-1-1.vturb.net:9000. DB::Exception: std::__1::__fs::filesystem::filesystem_error: filesystem error: in file_size: No such file or directory ["/var/lib/clickhouse/store/927/9274f8f4-22d6-4001-82a7-bedeeaa8c148/202408_304206_304206_0/by_created_v1.proj/data.cmrk3"]. (STD_EXCEPTION) (version 24.1.5.6 (official build))
, server ClickHouseNode [uri=http://[REDACTED]:8123/default, options={socket_timeout=300000}]@2112327647

Code: 1001. DB::Exception: Received from [REDACTED]:9000. DB::Exception: std::__1::__fs::filesystem::filesystem_error: filesystem error: in file_size: No such file or directory ["/var/lib/clickhouse/store/927/9274f8f4-22d6-4001-82a7-bedeeaa8c148/202408_304206_304206_0/by_created_v1.proj/data.cmrk3"]. (STD_EXCEPTION) (version 24.1.5.6 (official build))
```

**How to reproduce**

Example of INSERT:

```sql
insert into tmp.etl_sessions
      WITH  
        CAST('{{ inputs.data.start_time }}', 'DateTime64') as endDate,
        toStartOfDay(endDate) as startDate
      SELECT  company_id,
              video_id,
              player_id,
              any(device_id) as device_id,
              session_id,
              '{{ outputs.uuid.vars.id }}' as process_key
      FROM analytics2.sessions
      WHERE bot = 0
        AND created_at BETWEEN date_sub(HOUR, 8, startDate) AND date_add(HOUR, 3, endDate)
      GROUP BY company_id, player_id, video_id, session_id
      SETTINGS insert_distributed_sync = 1
```

```sql
INSERT INTO analytics2.events_hourly (company_id,
                                     video_id,
                                     player_id,
                                     date_key,
                                     event,
                                     random,
                                     total,
                                     total_uniq_sessions,
                                     total_uniq_device)
      WITH
      CAST('{{ outputs.uuid.vars.id }}', 'String') as pKey,
      CAST('{{ inputs.data.start_time }}', 'DateTime64') as startDate,
      CAST('{{ inputs.data.end_time }}', 'DateTime64') as endDate,
      toStartOfDay(startDate) as startDay,
      toStartOfHour(startDate) as startHour,
      randCanonical() as randomNumber,
      current_result as (SELECT
                            company_id,
                            video_id,
                            player_id,
                            event,
                            sum(total) as total,
                            sum(total_uniq_sessions) as total_uniq_sessions,
                            sum(total_uniq_device) as total_uniq_device
                        FROM analytics2.events_hourly final
                        WHERE date_key between startDay and endDate
                        group by 
                              company_id,
                              video_id,
                              player_id,
                              event),
      events_process as (SELECT e.company_id,
                            e.video_id,
                            e.player_id,
                            e.event                   as event,
                            count()                   as total,
                            uniqExact(fs.session_id)  as total_uniq_sessions,
                            uniqExact(fs.device_id)   as total_uniq_device
                    FROM analytics2.events as e
                    GLOBAL INNER JOIN (SELECT * FROM tmp.etl_sessions WHERE process_key = pKey) as fs
                                ON fs.session_id = e.session_id
                                   AND fs.company_id = e.company_id
                                   AND fs.player_id = e.player_id
                    WHERE e.created_at BETWEEN startDay and endDate
                    GROUP BY e.company_id,
                              e.video_id,
                              e.player_id,
                              event)
      select
            company_id,
            video_id,
            player_id,
            startHour as date_key,
            event,
            randomNumber as random,
            sum(t1.total) - ifNull(sum(t2.total), 0) as total,
            sum(t1.total_uniq_sessions) - ifNull(sum(t2.total_uniq_sessions), 0) as total_uniq_sessions,
            sum(t1.total_uniq_device) - ifNull(sum(t2.total_uniq_device), 0) as total_uniq_device
      from
            events_process as t1
      left join current_result as t2 on
            t1.company_id = t2.company_id
            and t1.video_id = t2.video_id
            and t1.player_id = t2.player_id
            and t1.event = t2.event
      group by
            company_id,
            video_id,
            player_id,
            event
      SETTINGS insert_distributed_sync = 1
```

**Tables Structure**
```sql
create table events_hourly_local
(
    company_id          String,
    video_id            String,
    player_id           String,
    date_key            DateTime,
    event               LowCardinality(String),
    random              Float64,
    total               Int64,
    total_uniq_sessions Int64,
    total_uniq_device   Int64
)
    engine = ReplicatedSummingMergeTree('/clickhouse/tables/{uuid}/{shard}', '{replica}')
        PARTITION BY toYYYYMM(date_key)
        PRIMARY KEY (company_id, event, date_key, player_id, video_id)
        ORDER BY (company_id, event, date_key, player_id, video_id)
        SETTINGS index_granularity = 32768, storage_policy = 's3_tiered', min_bytes_for_wide_part = 1048576000, min_rows_for_wide_part = 1000000;

create table events_hourly
(
    company_id          String,
    video_id            String,
    player_id           String,
    date_key            DateTime,
    event               LowCardinality(String),
    random              Float64,
    total               Int64,
    total_uniq_sessions Int64,
    total_uniq_device   Int64
)
    engine = Distributed('standard', 'analytics2', 'events_hourly_local', sipHash64(player_id));
```
    
**Expected behavior**

- Expected to be able to either insert / query without the DB failing to read from projections

**Additional context**

- We run with Kestra a consolidation job which each 5 minutes aggregates and inserts data into the event_hourly table, but, this problem of execution and failing to read the `data.bin` file from a .proj folder isn't exclusive on this process, our normal systems that query the clickhouse also get this exception and the weird part for me is that it's increasing the number of times this happens. We are basically having to write retries for all operations and the subsequent call would work but I'm pretty sure we should not be having this error, before moving to a cluster we didn't have this kind of problem.

PS: I'm not the creator of those tables and projections so any guidance here is appreciated since the creator is on vacation and possibly there's additional info that I'm just not sure how to share here.

**Comment by dakenblack** [comment]
Has there been any update on this, I believe I am running into the same issue. 

I am running a local instance through docker on windows (WSL2 backend).

Happy to provide any additional information, when I get time I'll try to create a reproducible example.

Cheers,

**Comment by dakenblack** [comment]
Hi,

I would like to give an update on my experiments.

## TLDR

I get the following error when trying to insert data into a local clickhouse server that is running in WSL (NOT Docker) but storing data in D drive. The server seems to work if it is pointing to C drive.

```
2025.04.02 16:48:10.403332 [ 7098 ] {369ea273-6fc5-4adf-845b-f46d8e817f97} <Debug> WriteBufferFromHTTPServerResponse: Write buffer has been canceled with an error. Proper HTTP error code and headers have been send to the client. HTTP code: HTTP_INTERNAL_SERVER_ERROR, message: <Internal Server Error>, error code: 1001, message: <std::exception. Code: 1001, type: std::__1::filesystem::filesystem_error, e.what() = filesystem error: in rename: Permission denied ["./store/31b/31b2c938-f08d-4e81-a7aa-9b3a7ad475cd/tmp_insert_all_1_1_0/"] ["./store/31b/31b2c938-f08d-4e81-a7aa-9b3a7ad475cd/all_1_1_0/"]
Cannot print extra info for Poco::Exception (version 25.4.1.1604 (official build))>, use compression: true, data has been send through buffers: false, compression discarded data: 0, discarded data: 0
```

## Steps

1. log in to Ubuntu WSL instance
2. Navigate to `/mnt/d` create clickhouse folder `mkdir Clickhouse` and cd `cd Clickhouse`
3. Download and run server `curl https://clickhouse.com | sh` + `./clickhouse server`
4. Elsewhere connect to server and run the following commands

```
CREATE TABLE TEST (
    "A" Float64,
    "B" Float64,
) ENGINE = MergeTree()
ORDER BY
    tuple();
    
INSERT INTO
    TEST
VALUES
    (10.0, 11.0);
```

I have tried step 4 a number of times with different clients:

1. ibis (python)
2. clickhouse_connect (python)
3. clickhouse client (second wsl terminal)
4. DBeaver (Tool for DB management)

## Attempted solutions

See references for where I got the ideas, but things I've tried are

- Changing ownership and permissions (I've also 777'd the /mnt/d drive)
- I've unmounted and mounted with `-o metadata`
- For good measure I've updated the `/etc/wsl.conf` with the following

```
[automount]
enabled = true
options = "metadata,umask=22,fmask=11"
```

All of the above (and combinations) of it has not worked.

If I create the folder inside the `~/clickhouse` folder (i.e. no mounting) then all the commands work. However, this does not work for me as I would like to use an alternative drive as I expect to store a lot of data on this server.

## References

I have read the following posts 

1. https://stackoverflow.com/a/65385081
2. https://stackoverflow.com/a/46967182

Any help would be greatly appreciated!

Cheers,
Jabez
